### PR TITLE
feat: name fields for definitionfile

### PR DESCRIPTION
**Pull request by soridalac**
### What does this PR do?
create name feature in definition-file for apex and group option fields 
### What issues does this PR fix or reference?
https://github.com/forcedotcom/cli/issues/3008#issuecomment-2342416748
@W-16757770@

### CODE REVIEW

File: messages/create.sandbox.md [hunk 1]
```diff
@@ -123,3 +123,19 @@ The poll interval (%d seconds) can't be larger that wait (%d in seconds)
...
 
 Could not find the clone sandbox name "%s" in the target org.
+
+# error.userGroupQueryFailed
+
+Could not query user group.
+
+# error.bothApexClassIdAndNameProvided
+
+Could not specify both apexClassId and apexClassName.
```
Comment (cristiand391) [inline messages/create.sandbox.md:133]:
```suggestion
You cannot specify both `apexClassId` and `apexClassName` in the definition file at the same time.
```

File: messages/create.sandbox.md [hunk 2]
```diff
@@ -123,3 +123,19 @@ The poll interval (%d seconds) can't be larger that wait (%d in seconds)
...
+
+# error.bothUserGroupIdAndNameProvided
+
+Could not specify both userGroupId and userClassName.
```
Comment (cristiand391) [inline messages/create.sandbox.md:137]:
```suggestion
You cannot specify both `ActivationUserGroupId` and `ActivationUserGroupName` in the definition file at the same time.
```

File: messages/create.sandbox.md [hunk 3]
```diff
@@ -123,3 +123,19 @@ The poll interval (%d seconds) can't be larger that wait (%d in seconds)
...
+
+# error.apexClassQueryFailed
+
+Could not query apex class.
```
Comment (cristiand391) [inline messages/create.sandbox.md:141]: This error might be confusing to users (they didn't wrote the definition file, just trying to create a sandbox). Which class the msg refers to? where is this class defined?

Maybe something like this?
```
Unable to find the ID of Apex class <className> defined in the definition file.
```

File: src/commands/org/create/sandbox.ts [hunk 4]
```diff
@@ -253,4 +279,30 @@ export default class CreateSandbox extends SandboxCommandBase<SandboxCommandResp
...
+      throw messages.createError('error.apexClassQueryFailed');
+    }
+  }
+
+  private async getUserGroupId(groupName: string): Promise<string | undefined> {
+    try {
+      const result = (
+        await this.flags['target-org']
+          .getConnection()
+          .singleRecordQuery(`SELECT id FROM Group WHERE NAME = '${groupName}' AND Type = 'Regular'`)
```
Comment (cristiand391) [inline src/commands/org/create/sandbox.ts:301]: Why does it only looks for regular user groups (`AND Type = 'Regular'`)?

File: messages/create.sandbox.md [hunk 5]
```diff
@@ -123,3 +123,19 @@ The poll interval (%d seconds) can't be larger that wait (%d in seconds)
 # error.noCloneSource
 
 Could not find the clone sandbox name "%s" in the target org.
+
+# error.userGroupQueryFailed
+
+Could not query user group.
```
Comment (cristiand391) [inline messages/create.sandbox.md:129]: ditto https://github.com/salesforcecli/plugin-org/pull/1221/files#r1773460127

File: src/commands/org/create/sandbox.ts [hunk 6]
```diff
@@ -253,4 +279,30 @@ export default class CreateSandbox extends SandboxCommandBase<SandboxCommandResp
...
+          .getConnection()
+          .singleRecordQuery(`SELECT Id FROM ApexClass WHERE Name = '${className}'`)
+      ).Id;
+      return result;
+    } catch (err) {
+      throw messages.createError('error.apexClassQueryFailed');
+    }
+  }
+
+  private async getUserGroupId(groupName: string): Promise<string | undefined> {
```
Comment (cristiand391) [inline src/commands/org/create/sandbox.ts:296]:
```suggestion
  private async getUserGroupIdByName(groupName: string): Promise<string | undefined> {
```

File: src/shared/sandboxRequest.ts [hunk 7]
```diff
@@ -33,13 +33,28 @@ export async function createSandboxRequest(
   definitionFile: string | undefined,
   logger?: Logger | undefined,
   properties?: Record<string, string | undefined>
-): Promise<{ sandboxReq: SandboxRequest; srcSandboxName: string }>;
+): Promise<{
+  sandboxReq: SandboxRequest & {
+    ApexClassName: string | undefined;
+    ApexClassId: string | undefined;
+    ActivationUserGroupName: string | undefined;
+    ActivationUserGroupId: string | undefined;
```
Comment (cristiand391) [inline src/shared/sandboxRequest.ts:41]: Of these 4 new fields, `ApexClassId` and `ActivationUserGroupId` are real properties in the sandbox request, let's add them here instead:
https://github.com/forcedotcom/sfdx-core/blob/4580a1c323379292e6da1ddb0e00458782ca1b51/src/org/org.ts#L128

`SandboxRequest` represents the payload for a sandbox create/refresh request so we don't want to add the apexclass/activationusergroup `Name` fields there (they only exists in the definition file that we handle)

**Review by cristiand391 — CHANGES_REQUESTED** [review]
looks good over all, left some questions and requested some small changes

File: messages/create.sandbox.md (hunk 5 above)
Comment (jshackell-sfdc) [reply messages/create.sandbox.md:129]: I like Cristian's suggestion below, so how about this:

`Unable to find the ID of the activation user group <activationGroupName> that you defined in the definition file. `
I assume that in your code, <activationGroupName> will actually be something like `<%s>`, right?

File: messages/create.sandbox.md (hunk 1 above)
Comment (jshackell-sfdc) [reply messages/create.sandbox.md:133]:
```suggestion
You can't specify both `apexClassId` and `apexClassName` in the definition file at the same time.
```

Comment (jshackell-sfdc) [reply messages/create.sandbox.md:133]: Love Cristian's suggestion, I just have a small tweak to it!

File: messages/create.sandbox.md (hunk 2 above)
Comment (jshackell-sfdc) [reply messages/create.sandbox.md:137]:
```suggestion
You can't specify both `ActivationUserGroupId` and `ActivationUserGroupName` in the definition file at the same time.
```

Comment (jshackell-sfdc) [reply messages/create.sandbox.md:137]: Ditto, I have a small tweak to Cristian's suggestion.

File: messages/create.sandbox.md (hunk 3 above)
Comment (jshackell-sfdc) [reply messages/create.sandbox.md:141]: Slight edit to Cristian's suggestion:

`Unable to find the ID of the Apex class <className> that's defined in the definition file.`

**Review by cristiand391 — APPROVED** [review]

**Comment by cristiand391** [comment]
**QA notes**:

setup:
checked out this PR, `yarn compile` then linked it with `sf plugins link`

### `org create sandbox`

✅ throws when specifying both `ApexClassId` and `ApexClassName`
✅ throws when specifying both `ActivationUserGroupId`and `ActivationUserGroupName`

⚠️ NIT: the validation happens after the sandbox name is generated so we always get this warning, let's do the validation early 👍🏼 
Screenshot 2024-09-25 at 10 54 19 AM

✅ throws if can't find ID of `ApexClassName`

⚠️ NIT: 
Screenshot 2024-09-25 at 11 00 36 AM

✅ throws if can't find ID of `ActivationUserGroupName`

Screenshot 2024-09-25 at 11 01 42 AM

✅ resolves `ApexClassName` to class ID
def.file 
```json
{
  "description": "testing testing",
  "ApexClassName": "PropertyController"
}
```
...
```
=== Config Sandbox Request

 Field       Value
 ─────────── ──────────────────
 Description testing testing
 SandboxName sbx1hxq4ir
 LicenseType Developer
 ApexClassId 01p4p00000AryNlAAJ
```
Screenshot 2024-09-25 at 11 27 19 AM

✅ resolves `ActivationUserGroupName` to user group ID
✅ `ApexClassName` works without `ActivationUserGroupName`
✅ `ActivationUserGroupName` works without `ApexClassName`

### `org refresh sandbox`
❌ throws when specifying both `ApexClassId` and `ApexClassName`
❌ throws when specifying both `ActivationUserGroupId`and `ActivationUserGroupName`
❌ throws if can't find ID of `ApexClassName`
❌ throws if can't find ID of `ActivationUserGroupName`
❌ resolves `ApexClassName` to class ID
❌ resolves `ActivationUserGroupName` to user group ID
❌ `ApexClassName` works without `ActivationUserGroupName`
❌ `ActivationUserGroupName` works without `ApexClassName`

**Review by cristiand391 — APPROVED** [review]

**Comment by cristiand391** [comment]
**QA update**

> ⚠️ NIT: the validation happens after the sandbox name is generated so we always get this warning, let's do the validation early 👍🏼

✅ fixed
Screenshot 2024-09-25 at 5 05 20 PM

Screenshot 2024-09-25 at 5 06 46 PM

### `org refresh sandbox`
✅  throws when specifying both ApexClassId and ApexClassName
✅  throws when specifying both ActivationUserGroupIdand ActivationUserGroupName
✅  throws if can't find ID of ApexClassName
✅  throws if can't find ID of ActivationUserGroupName
✅  resolves ApexClassName to class ID (no apex class -> class assigned, old class -> new class)
✅  resolves ActivationUserGroupName to user group ID (no group -> group assigned, old group -> new group)
✅  ApexClassName works without ActivationUserGroupName
✅  ActivationUserGroupName works without ApexClassName

✅ props can start with/without capital letter, we handle casing when reading the def. file

**Pull request merged**